### DISCUSSION TITLE
How to make a package installed-by-request

### DISCUSSION THREAD

**Discussion by ghost** [Homebrew/discussions#4043] [Everyday usage] [answered]
The scenario is as follows:

1. I install package A. A depends on B, so B gets installed and brew knows (somehow) that it is installed as a dependency only. It has to know that because B is subject to future uninstalling with `brew autoremove`.
2. I do some personal development using B. This happens all the time if B is some library package.
3. I decide to uninstall A. Now B is really vulnerable to `autoremove` but I am using it directly, so I want to keep it! Hence:
4. I need a way to "forget" that B is just a dependency and treat it as if it had been installed directly by my request.

How to do that? At least some similar packaging systems have a command like this. For example, [opam](https://opam.ocaml.org/doc/man/opam-install.html) has `opam install --set-root`.

**Comment by SMillerDev** [comment]
Have you tried `brew reinstall`?

**Reply by ghost** [reply]
No, although it was one of the possibilities I had looked into before I asked. Two reasons why not:

1. The help output doesn't really say it would have this effect, so it would be shooting in the dark which I hate doing.
2. The help output *does* say that it uninstalls as an intermediate step. Now, in my immediate situation this would not be a problem, because the program I developed with the library is only used by me from the command line. But in general, it could be in cron jobs, or even used by others. All such usage would be broken while running the command.

**Comment by Bo98** [comment, marked as answer]
`brew install` will do this. If it is already installed, it will set it to installed on request.

A couple things about it to be aware:

* If it's already installed and up-to-date, it will print a warning that it is "already installed", but it still will have updated the setting successfully. If this message is a problem you can add `--quiet`.
* If it's already installed but _not_ up-to-date, it will `brew upgrade` that formula too so that the latest version is installed. If this is a problem, you can do `HOMEBREW_NO_INSTALL_UPGRADE=1 brew install` instead.

**Reply by ghost** [reply]
Thank you! That should work for me.
